examples/winit_multithread.rs

```diff
@@ -59,41 +63,55 @@ mod ex {
     pub(super) fn entry() {
         let event_loop = EventLoop::new().unwrap();
 
-        let app = winit_app::WinitAppBuilder::with_init(|elwt| {
-            let attributes = Window::default_attributes();
-            #[cfg(target_arch = "wasm32")]
-            let attributes =
-                winit::platform::web::WindowAttributesExtWebSys::with_append(attributes, true);
-            let window = Arc::new(elwt.create_window(attributes).unwrap());
-
-            let context = softbuffer::Context::new(window.clone()).unwrap();
-            let surface = {
-                println!("making surface...");
-                let surface = softbuffer::Surface::new(&context, window.clone()).unwrap();
-                Arc::new(Mutex::new(surface))
-            };
-
-            // Spawn a thread to handle rendering.
-            let (start_render, do_render) = mpsc::channel();
-            let (render_done, finish_render) = mpsc::channel();
-            println!("starting thread...");
-            std::thread::spawn({
-                let window = window.clone();
-                let surface = surface.clone();
-                move || render_thread(window, surface, do_render, render_done)
-            });
-
-            (window, surface, start_render, finish_render)
-        })
-        .with_event_handler(|state, event, elwt| {
-            let (window, _surface, start_render, finish_render) = state;
+        let app = winit_app::WinitAppBuilder::with_init(
+            |elwt| {
+                let attributes = Window::default_attributes();
+                #[cfg(target_arch = "wasm32")]
+                let attributes =
+                    winit::platform::web::WindowAttributesExtWebSys::with_append(attributes, true);
+                let window = Arc::new(elwt.create_window(attributes).unwrap());
+
+                let context = softbuffer::Context::new(window.clone()).unwrap();
+
+                (window, context)
+            },
+            |_elwt, (window, context)| {
+                let surface = {
+                    println!("making surface...");
+                    let surface = softbuffer::Surface::new(context, window.clone()).unwrap();
+                    Arc::new(Mutex::new(surface))
+                };
+
+                // TODO: This sample will now spawn a new thread every time the surface is
+                // (re)created.  We should instead start it only once, and send an `Arc`ed clone of
+                // the surface through do_render?
+
+                // Spawn a thread to handle rendering.
+                let (start_render, do_render) = mpsc::channel();
+                let (render_done, finish_render) = mpsc::channel();
+                println!("starting thread...");
+                std::thread::spawn({
+                    let window = window.clone();
+                    let surface = surface.clone();
+                    move || render_thread(window, surface, do_render, render_done)
+                });
```

Comment:
Thoughts?

Reply:
I actually think its fine like this.
The thread is destroyed on `suspended()` anyway, maybe we should leave a comment stating as much.

Probably update this line as well:
https://github.com/rust-windowing/softbuffer/blob/592c49e4876979bc09b8f5002edcb9b345f6c9c5/examples/winit_multithread.rs#L29

Reply:
Okay, I've updated this a bit but would like to test it, including the idea to keep one thread alive for the context and sending it the `Surface` as part of the app (not surface) state.

Reply:
Forgot to mention that I still wanted to test this on an actual Android phone, both making sure that the current logic works and the idea I had about persistent threads.

Reply:
```
08-03 11:15:00.827 15549 15549 D android_activity::act..: NativeWindowCreated: 0xb400007229e9ff70 -- 0xb400007369e7ef10
08-03 11:15:00.827 15549 15549 D android_activity::act..: NativeWindowResized: 0xb400007229e9ff70 -- 0xb400007369e7ef10
08-03 11:15:00.828 15549 15598 I RustStdoutStderr: making surface...
08-03 11:15:00.828 15549 15598 I RustStdoutStderr: starting thread...
08-03 11:15:00.829 15549 15598 I RustStdoutStderr: waiting for render...
08-03 11:15:00.829 15549 15549 D android_activity::act..: ContentRectChanged: 0xb400007229e9ff70 -- 0x7fded9b708
08-03 11:15:00.831 15549 15549 D android_activity::act..: NativeWindowRedrawNeeded: 0xb400007229e9ff70 -- 0xb400007369e7ef10
08-03 11:15:00.832 15549 15598 I RustStdoutStderr: got size: PhysicalSize { width: 1096, height: 2560 }
08-03 11:15:00.832 15549 15598 I RustStdoutStderr: resizing...
08-03 11:15:00.838 15549 15603 I AdrenoUtils: <ReadGpuID:408>: Reading chip ID through GSL
08-03 11:15:00.851 15549 15549 D android_activity::act..: WindowFocusChanged: 0xb400007229e9ff70 -- 1
08-03 11:15:01.087 15549 15598 I RustStdoutStderr: presenting...
08-03 11:15:01.087 15549 15598 I RustStdoutStderr: waiting for render...
08-03 11:15:01.106 15549 15575 I ithread_android: Compiler allocated 4265KB to compile void android.widget.TextView.<init>(android.content.Context, android.util.AttributeSet, int, int)
08-03 11:15:01.165 15549 15575 I ithread_android: Compiler allocated 4683KB to compile void android.view.ViewRootImpl.performTraversals()
08-03 11:15:03.577 15549 15549 D android_activity::act..: Pause: 0xb400007229e9ff70
08-03 11:15:03.611 15549 15549 D android_activity::act..: NativeWindowDestroyed: 0xb400007229e9ff70 -- 0xb400007369e7ef10
08-03 11:15:03.612 15549 15598 I RustStdoutStderr: Main thread destroyed
```
Works as it should on Android!

(note that I added `Main thread destroyed` in `do_render.recv().is_err()` even though that is not really what it does currently, but it shows that destruction of the surface state and the relevant `mpsc` is happening correctly.)

Reply:
> (note that I added `Main thread destroyed` in `do_render.recv().is_err()` even though that is not really what it does currently, but it shows that destruction of the surface state and the relevant `mpsc` is happening correctly.)

You corrected it to:
https://github.com/rust-windowing/softbuffer/blob/eb6be27e0284a93ec613cdc95dd6bc4aee724fdb/examples/winit_multithread.rs#L29

So looks good, right?

Reply:
I meant in the example log above.

I've ultimately rectified it in both commits in the history, as I ended up pushing this extra log line for testing.